File: js_algo/merge_sort.js
function merge(arr, startLeft, endLeft, end){
    sizeLeft = endLeft - startLeft + 1
    sizeRight = end - endLeft
    startRight = endLeft + 1

    arrLeft = []
    arrRight = []

    //Copy array into temporay storage
    for(let i = 0; i < sizeLeft; i++){
        arrLeft[i] = arr[startLeft + i]
    }
    arrLeft[sizeLeft] = Number.MAX_SAFE_INTEGER

    for(let i = 0; i < sizeRight; i++){
        arrRight[i] = arr[startRight + i]
    }
    arrRight[sizeRight] = Number.MAX_SAFE_INTEGER


    let indexLeft = 0;
    let indexRight = 0;
    for(let i = startLeft; i <= end; i++ ){
        if(arrLeft[indexLeft] <= arrRight[indexRight]){
            arr[i] = arrLeft[indexLeft]
            indexLeft += 1;
        }else{
            arr[i] = arrRight[indexRight]
            indexRight += 1;
        }
    }

    //console.log("after meger", arr)
}

function mergeSort(arr, start, end){
    if(start < end){
        let endLeft = Math.floor( (start + end ) / 2)
        mergeSort(arr, start, endLeft)
        mergeSort(arr, endLeft + 1, end)
        merge(arr, start, endLeft, end)
    }
}

function mergeSortWrapper(arr){
    let size = arr.length
    mergeSort(arr, 0, size -1)
}


arr = [43, 24, 45, 27, 35, 47, 22, 48]
//merge(arr, 0,3, 7)
//console.log(arr)
mergeSortWrapper(arr)

console.log("run merge", arr)

File: merge_sort.py
#starting from the merge function
# 
import sys

def merge(array, startLeft, endLeft, end):
    sizeLeft = endLeft - startLeft + 1
    sizeRight = end - endLeft
    startRight = endLeft + 1

    # create a new temporary storage for size n
    arrLeft = [None] * (sizeLeft + 1)
    arrRight = [None] * (sizeRight + 1)

    #copy the array into the temporary storage
    for i in range(sizeLeft):
        arrLeft[i] = array[startLeft + i]
    arrLeft[sizeLeft] = sys.maxsize # set the sentinel

    for i in range(sizeRight):
        arrRight[i] = array[startRight + i]
    arrRight[sizeRight] = sys.maxsize

    #Merge the array by taking the smallest first
    indexLeft = 0
    indexRight = 0
    for i in range(startLeft, end + 1):
        if arrLeft[indexLeft] <= arrRight[indexRight]:
            array[i] = arrLeft[indexLeft]
            indexLeft = indexLeft + 1
        else:
            array[i] = arrRight[indexRight]
            indexRight = indexRight + 1
            
    pass

#Develop the merge sort alog
def mergeSort(array, start, end):
    if(start < end):
        endLeft = (start + end) // 2
        mergeSort(array, start, endLeft)
        mergeSort(array, endLeft + 1, end)
        merge(array, start, endLeft, end)

#Develop the user Interface for merge sort
def mergeSortWrapper(array):
    size = len(array)
    mergeSort(array, 0, size - 1)
   

arr = [43, 24, 45, 27, 35, 47, 22, 48]

#merge(arr, 0, 3, 7)
mergeSortWrapper(arr)
print(arr)
  